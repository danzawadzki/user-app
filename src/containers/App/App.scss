@import '../../assets/styles/minor';

/** App - component styles */
.App {
	width: 50rem;
	background-color: white;
	border-radius: $radius--default;
	box-shadow: inset 0 11.5rem 0 -2rem $brand--main, $shadow--default;
	padding: 3.6rem 1.7rem;
	box-sizing: border-box;
	max-height: 80rem;
	min-height: 80rem;
	overflow: hidden;
	display: flex;
	flex-direction: column;

	& .userProfile__counters {
		display: flex;
		align-items: center;
		justify-content: space-between;
		margin-top: 2rem;
		margin-bottom: 1rem;
	}

	#UserProfile {
		flex-grow: 0;
	}
	#Comments {
		flex-grow: 1;
		overflow: scroll;
	}
}

@media (max-width: 32rem) {
	.App {
		& .userProfile__counters {
			flex-direction: column;
		}
	}
}
