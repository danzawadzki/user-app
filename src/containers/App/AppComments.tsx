import Accordion from '../Accordion/Accordion';
import CommentsList from '../../components/CommentsList/CommentsList';
import MessageInput from '../../components/MessageInput/MessageInput';
import Module from '../../components/Module/Module';
import React from 'react';
import { IComment } from '../../components/Comment/Comment';

export interface IAppComments {
	/** Comments array */
	comments?: Array<IComment>;
	/** Value of comment form */
	commentForm?: string;
	/** Handler to a onChange event. */
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
	/** Handler to a onSubmit event. */
	onSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
}

/**
 * AppComments component. Contains accordion component and message input to place new posts.
 * @constructor
 */
const AppComments = ({ comments, commentForm, onChange, onSubmit }: IAppComments) => (
	<Module id="Comments">
		<Accordion
			heading={{
				on: `Hide comments (${comments ? comments.length : 0})`,
				off: `Show comments (${comments ? comments.length : 0})`
			}}>
			<CommentsList comments={comments} />
		</Accordion>
		<MessageInput
			id="commentForm"
			value={commentForm}
			placeholder="Add a comment..."
			onChange={onChange}
			onSubmit={onSubmit}
		/>
	</Module>
);

export default AppComments;
