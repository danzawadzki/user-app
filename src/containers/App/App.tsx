import React, { Component } from 'react';
import './App.scss';
import { connect } from 'react-redux';
import Module from '../../components/Module/Module';
import { IAvatarSrc } from '../../components/Avatar/Avatar';
import { ICounter } from '../../components/Counter/Counter';
import { addComment, fetchComments } from '../../actions/comments.actions';
import { fetchUser, incrementUserCounter } from '../../actions/user.actions';
import CommentsFeed from '../CommentsFeed/CommentsFeed';
import UserProfile from '../UserProfile/UserProfile';

export interface IAppUser {
	/** User name */
	name: string;
	/** User location */
	location?: string;
	/** User avatar */
	avatar?: IAvatarSrc;
	/** Comment content body */
	counters?: ICounter[];
}

export interface IAppState {
	/** Controlled input value with comment to post */
	commentForm?: string;
}

class App extends Component<any, IAppState> {
	state = {
		commentForm: ''
	};
	/**
	 * Handler to form onSubmit event.
	 * @param e
	 */
	handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();
		// Spreading state and props
		const { commentForm } = this.state;
		const { user, addComment } = this.props;

		// Validate if comment contains body
		if (!commentForm) {
			return console.log('Type something!');
		}
		addComment({ avatar: user.data.avatar, name: user.data.name }, commentForm);
		// Setting up state
		this.setState({
			commentForm: ''
		});
	};

	/**
	 * Fetching commens and user profile info on component mount
	 */
	async componentDidMount() {
		this.props.fetchUser();
		this.props.fetchComments();
	}

	/**
	 * Handler to input onChange event.
	 * @param e
	 */
	handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		e.persist();
		this.setState(prevState => ({
			...prevState,
			[e.target.id]: e.target.value
		}));
	};

	render() {
		return (
			<div className="App">
				<Module id="UserProfile">
					<UserProfile
						handleIncrement={this.props.incrementUserCounter}
						isLoading={this.props.user.isLoading}
						user={this.props.user.data}
					/>
				</Module>
				<Module id="Comments">
					<CommentsFeed
						isLoading={this.props.comments.isLoading}
						commentForm={this.state.commentForm}
						comments={this.props.comments.data}
						onChange={this.handleChange}
						onSubmit={this.handleSubmit}
					/>
				</Module>
			</div>
		);
	}
}

/**
 * Mapping redux state to props.
 */
const mapStateToProps = (state: any) => {
	return {
		comments: state.comments,
		user: state.user
	};
};

export default connect(
	mapStateToProps,
	{
		fetchUser,
		incrementUserCounter,
		fetchComments,
		addComment
	}
)(App);
