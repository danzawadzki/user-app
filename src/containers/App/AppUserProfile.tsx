import * as React from 'react';
import './App.scss';
import UserCard from '../../components/UserCard/UserCard';
import CountersList from '../../components/CountersList/CountersList';
import Button from '../../components/Button/Button';
import { IAppUser } from './App';
import WithLoader from '../../components/WithLoader/WithLoader';

export interface IAppUserProfile {
	user: IAppUser;
	handleIncrement: (label: string) => void;
}

const AppUserProfile = ({ user, handleIncrement }: IAppUserProfile) => {
	return (
		<>
			<UserCard
				name={user.name}
				location={user.location}
				avatar={user.avatar}
				onHeartClick={() => handleIncrement('likes')}
			/>
			<div className="userProfile__counters">
				<CountersList counters={user.counters} />
				<Button id="follow" onClick={() => handleIncrement('followers')}>
					Follow
				</Button>
			</div>
		</>
	);
};

export default WithLoader(AppUserProfile);
