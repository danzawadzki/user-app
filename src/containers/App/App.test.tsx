import * as React from 'react';
import App from './App';
import { shallow } from 'enzyme';

describe('App container: ', () => {
	// // Shallow rendered component
	const component = shallow(<App />);

	it('should render match snapshot', () => {
		expect(component).toMatchSnapshot();
	});
});
