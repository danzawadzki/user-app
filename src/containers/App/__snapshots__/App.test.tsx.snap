// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App container:  should render match snapshot 1`] = `<Connect(App) />`;
